Array.index function genereated for array of sumtype references produces C error.

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.4 223b96a.1fc9e1a
**OS:** linux, "Manjaro Linux"

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
struct Element {
	AbstractNode
mut:
	name       string
	value      string
	attributes []&Attribute
}

struct Attribute {
	AbstractNode
mut:
	name  string
	value string
}

pub type Node = Attribute | Element

struct AbstractNode {
pub mut:
	child_nodes []&Node
}

pub fn (mut n AbstractNode) append_child(child &Node) {
	n.child_nodes << child
}

pub fn (n &AbstractNode) has_child(child &Node) int {
	return n.child_nodes.index(child)
}

fn main() {
	mut parent := &Element{
		name: 'parent'
	}
	mut child := &Element{
		name: 'child'
	}
	parent.append_child(child)

	println(parent.has_child(child))
	dump(parent)
}
```


**What did you expect to see?**
Expected code to compile and the `has_child` function to return an integer.
 
**What did you see instead?**
The code doesn't compile because the generated C code is invalid.
```
/usr/lib64/crt1.o: error: Invalid relocation entry [15] '.rela.debug_info' @ 000004f5
tcc: error: file 'crt1.o' not found
/tmp/v_1000/test.9025346761131596870.tmp.c:2213: error: cannot convert 'struct main__Node *' to 'struct main__Node'
```
```c
static int Array_dom__Node_ptr_index(Array_dom__Node_ptr a, dom__Node* v) {
	dom__Node** pelem = a.data;
	for (int i = 0; i < a.len; ++i, ++pelem) {
		if (dom__Node_sumtype_eq(*pelem, v)) {
			return i;
		}
	}
	return -1;
}
```

--- a/vlib/v/gen/c/array.v
+++ b/vlib/v/gen/c/array.v
@@ -858,10 +858,18 @@ fn (mut g Gen) gen_array_index_methods() {
 			fn_builder.writeln('\t\tif (${ptr_typ}_struct_eq(*pelem, v)) {')
 		} else if elem_sym.kind == .interface_ {
 			ptr_typ := g.equality_fn(info.elem_type)
-			fn_builder.writeln('\t\tif (${ptr_typ}_interface_eq(*pelem, v)) {')
+			if info.elem_type.is_ptr() {
+				fn_builder.writeln('\t\tif (${ptr_typ}_interface_eq(**pelem, *v)) {')
+			} else {
+				fn_builder.writeln('\t\tif (${ptr_typ}_interface_eq(*pelem, v)) {')
+			}
 		} else if elem_sym.kind == .sum_type {
 			ptr_typ := g.equality_fn(info.elem_type)
-			fn_builder.writeln('\t\tif (${ptr_typ}_sumtype_eq(*pelem, v)) {')
+			if info.elem_type.is_ptr() {
+				fn_builder.writeln('\t\tif (${ptr_typ}_sumtype_eq(**pelem, *v)) {')
+			} else {
+				fn_builder.writeln('\t\tif (${ptr_typ}_sumtype_eq(*pelem, v)) {')
+			}
 		} else if elem_sym.kind == .alias {
 			ptr_typ := g.equality_fn(info.elem_type)
 			fn_builder.writeln('\t\tif (${ptr_typ}_alias_eq(*pelem, v)) {')

--- a/vlib/v/tests/array_of_reference_sumtype_test.v
+++ b/vlib/v/tests/array_of_reference_sumtype_test.v
@@ -1,11 +1,14 @@
 struct Element {
 	AbstractNode
+mut:
 	name       string
+	value      string
 	attributes []&Attribute
 }
 
 struct Attribute {
 	AbstractNode
+mut:
 	name  string
 	value string
 }
@@ -21,7 +24,11 @@ pub fn (mut n AbstractNode) append_child(child &Node) {
 	n.child_nodes << child
 }
 
-fn test_array_of_reference_sumtype_append() {
+pub fn (n &AbstractNode) has_child(child &Node) int {
+	return n.child_nodes.index(child)
+}
+
+fn test_array_of_reference_sumtype() {
 	mut parent := &Element{
 		name: 'parent'
 	}
@@ -30,6 +37,9 @@ fn test_array_of_reference_sumtype_append() {
 	}
 	parent.append_child(child)
 
+	ret := parent.has_child(child)
+	println(ret)
 	dump(parent)
 	assert parent.child_nodes[0].name == 'child'
+	assert ret == 0
 }

